lib/ansible/modules/cloud/docker/docker_swarm_service.py

```diff
@@ -0,0 +1,905 @@
+        cspec['Mounts'] = []
+        for mount_config in self.mounts:
+            cspec['Mounts'].append({
+                'Target': mount_config['target'],
+                'Source': mount_config['source'],
+                'Type': mount_config['type'],
+                'ReadOnly': mount_config['readonly']
+            })
+        cspec['DNSConfig'] = {"Nameservers": self.dns,
+                              "Search": self.dns_search,
+                              "Options": self.dns_options}
+        cspec['Args'] = self.args
+        cspec['Env'] = self.env
+        cspec['TTY'] = self.tty
+        cspec['Hostname'] = self.hostname
+        cspec['Labels'] = self.container_labels
+        restart_policy = types.RestartPolicy(
+            condition=self.restart_policy,
+            delay=self.restart_policy_delay,
+            max_attempts=self.restart_policy_attempts,
+            window=self.restart_policy_window)
+        resources = {
+            'Limits': {
+                'NanoCPUs': int(self.limit_cpu * 1000000000),
+                'MemoryBytes': self.limit_memory * 1024 * 1024},
+            'Reservations': {
+                'NanoCPUs': int(self.reserve_cpu * 1000000000),
+                'MemoryBytes': self.reserve_memory * 1024 * 1024}}
+        task_template = types.TaskTemplate(
+            container_spec=cspec,
+            restart_policy=restart_policy,
+            placement=self.constraints,
+            resources=resources)
+        mode = {'Replicated': {'Replicas': self.replicas}}
+        if self.mode == 'global':
+            mode = {'Global': {}}
+
+        networks = []
+        for network_name in self.networks:
+            network_id = None
+            try:
+                network_id = filter(lambda n: n['name'] == network_name, docker_networks)[0]['id']
+            except:
+                pass
+            if network_id:
+                networks.append({'Target': network_id})
+            else:
+                raise Exception("no docker networks named: %s" % network_name)
+
+        endpoint_spec = {'Mode': self.endpoint_mode}
+        endpoint_spec['Ports'] = []
+        for port in self.publish:
+            endpoint_spec['Ports'].append({
+                'Protocol': port['protocol'],
+                'PublishMode': port['mode'],
+                'PublishedPort': int(port['published_port']),
+                'TargetPort': int(port['target_port'])})
+        return task_template, networks, endpoint_spec, mode, self.labels
+
+    def fail(self, msg):
+        self.parameters.client.module.fail_json(msg=msg)
+
+    @property
+    def exists(self):
+        return True if self.service else False
+
+
+class DockerServiceManager():
+    def get_networks_names_ids(self):
+        return [{'name': n['Name'], 'id': n['Id']} for n in self.client.networks()]
+
+    def get_service(self, name):
+        raw_data = self.client.services(filters={'name': name})
+        if len(raw_data) == 0:
```

Comment:
Seems like filter can return partial matches: `foo` would match for an existing `foo-bar` service, and would fail with thinking it needs to be updated (which also seems to be deprecated. Maybe it's worth verifying with this condition and adding a workaround:
```python
if len(raw_data) == 0 or not 'Name' in raw_data[0]['Spec'] or raw_data[0]['Spec']['Name'] != name:
```

Reply:
Do a pull request.

Reply:
There were so many migrations,I'm confused where is the original PR coming from. Is it https://github.com/dariko/ansible-modules-core/blob/swarm_service/cloud/docker/docker_swarm_service.py ?

Reply:
Yes

Reply:
Yes, it is.

Reply:
@kireevco sorry for the late reply. Gitub sure doesn't make it is to find the source branch of a PR :) I'm confident it's this one: https://github.com/dariko/ansible/tree/docker_swarm_service

Reply:
@kireevco  I was confident, and I was wrong. The source of this PR is https://github.com/dariko/ansible/tree/ansible-modules-core/pull/5547.
Sorry, but this PR have a messy and long history.